File: 7/1.py

def recur(curr, remaining, options):
    if not remaining:
        options.append(curr)
        return
    
    recur(curr * remaining[0], remaining[1:], options)
    recur(curr + remaining[0], remaining[1:], options)


def compute(nums):
    options = []

    recur(nums[0], nums[1:], options)

    return options


total = 0

while True:
    try:
        line = input().strip()
    except EOFError:
        break

    chunks = line.split(":")
    goal = int(chunks[0])

    rest = chunks[1].strip()

    nums = [int(num) for num in rest.split()]

    if goal in compute(nums):
        total += goal

print(total)

File: 7/1/1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func recur(curr int, remaining []int, goal int) bool {
	if len(remaining) == 0 {
		if curr == goal {
			return true
		}
		return false
	}

	return (recur(curr*remaining[0], remaining[1:], goal) ||
		recur(curr+remaining[0], remaining[1:], goal))
}

func main() {

	scanner := bufio.NewScanner(os.Stdin)

	ans := 0

	for scanner.Scan() {
		var goal int
		nums := []int{}

		line := scanner.Text()
		chunks := strings.Split(line, ":")

		fmt.Sscanf(chunks[0], "%v", &goal)

		for _, s := range strings.Split(chunks[1][1:], " ") {
			num, _ := strconv.Atoi(s)
			nums = append(nums, num)
		}

		if recur(nums[0], nums[1:], goal) {
			ans += goal
		}

	}

	fmt.Println(ans)
}
